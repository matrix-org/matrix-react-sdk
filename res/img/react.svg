<svg xmlns="http://www.w3.org/2000/svg" width="18" height="17" viewBox="0 0 18 17">
    <g fill="none" fill-rule="evenodd">
        <path fill="#2E2F32" fill-rule="nonzero" d="M13.6 14.5a7.875 7.875 0 1 1 2.183-3h-.818a7.125 7.125 0 1 0-2.62 3H13.6z"/>
        <g stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75">
            <path d="M11.5 13h6M14.5 10v6"/>
        </g>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width=".75" d="M5.5 10s1.125 1.5 3 1.5 3-1.5 3-1.5"/>
        <path stroke="#2E2F32" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6.25 6.25h.008M10.75 6.25h.008"/>
    </g>
</svg>
