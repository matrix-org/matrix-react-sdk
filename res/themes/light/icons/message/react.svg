<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><path d="M248 8C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 464c-119.1 0-216-96.9-216-216S128.9 40 248 40s216 96.9 216 216-96.9 216-216 216zm90.2-146.2C315.8 352.6 282.9 368 248 368s-67.8-15.4-90.2-42.2c-5.7-6.8-15.8-7.7-22.5-2-6.8 5.7-7.7 15.7-2 22.5C161.7 380.4 203.6 400 248 400s86.3-19.6 114.8-53.8c5.7-6.8 4.8-16.9-2-22.5-6.8-5.6-16.9-4.7-22.6 2.1zM168 240c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32zm160 0c17.7 0 32-14.3 32-32s-14.3-32-32-32-32 14.3-32 32 14.3 32 32 32z"/></svg>