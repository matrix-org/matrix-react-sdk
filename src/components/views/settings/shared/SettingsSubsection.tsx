/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classNames from "classnames";
import React, { HTMLAttributes } from "react";

import { SettingsSubsectionHeading } from "./SettingsSubsectionHeading";

export interface SettingsSubsectionProps extends HTMLAttributes<HTMLDivElement> {
    heading?: string | React.ReactNode;
    description?: string | React.ReactNode;
    children?: React.ReactNode;
    // when true content will be justify-items: stretch, which will make items within the section stretch to full width.
    stretchContent?: boolean;
}

export const SettingsSubsectionText: React.FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...rest }) => (
    <div {...rest} className="mx_SettingsSubsection_text">
        {children}
    </div>
);

export const SettingsSubsection: React.FC<SettingsSubsectionProps> = ({
    heading,
    description,
    children,
    stretchContent,
    ...rest
}) => (
    <div {...rest} className="mx_SettingsSubsection">
        {typeof heading === "string" ? <SettingsSubsectionHeading heading={heading} /> : <>{heading}</>}
        {!!description && (
            <div className="mx_SettingsSubsection_description">
                <SettingsSubsectionText>{description}</SettingsSubsectionText>
            </div>
        )}
        {!!children && (
            <div
                className={classNames("mx_SettingsSubsection_content", {
                    mx_SettingsSubsection_contentStretch: !!stretchContent,
                    mx_SettingsSubsection_noHeading: !heading && !description,
                })}
            >
                {children}
            </div>
        )}
    </div>
);

export default SettingsSubsection;
