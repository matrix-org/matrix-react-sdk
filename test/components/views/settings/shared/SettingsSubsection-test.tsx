/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { render } from "@testing-library/react";

import SettingsSubsection from "../../../../../src/components/views/settings/shared/SettingsSubsection";

describe("<SettingsSubsection />", () => {
    const defaultProps = {
        heading: "Test",
        children: <div>test settings content</div>,
    };
    const getComponent = (props = {}): React.ReactElement => <SettingsSubsection {...defaultProps} {...props} />;

    it("renders with plain text heading", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders with react element heading", () => {
        const heading = <h3>This is the heading</h3>;
        const { container } = render(getComponent({ heading }));
        expect(container).toMatchSnapshot();
    });

    it("renders without description", () => {
        const { container } = render(getComponent());
        expect(container).toMatchSnapshot();
    });

    it("renders with plain text description", () => {
        const { container } = render(getComponent({ description: "This describes the subsection" }));
        expect(container).toMatchSnapshot();
    });

    it("renders with react element description", () => {
        const description = (
            <p>
                This describes the section <a href="/#">link</a>
            </p>
        );
        const { container } = render(getComponent({ description }));
        expect(container).toMatchSnapshot();
    });
});
