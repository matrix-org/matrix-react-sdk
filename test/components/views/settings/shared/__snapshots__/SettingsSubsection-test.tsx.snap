// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsSubsection /> renders with plain text description 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Test
    </h3>
    <div
      class="mx_SettingsSubsection_description"
    >
      This describes the subsection
    </div>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div>
        test settings content
      </div>
    </div>
  </div>
</div>
`;

exports[`<SettingsSubsection /> renders with react element description 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Test
    </h3>
    <div
      class="mx_SettingsSubsection_description"
    >
      <p>
        This describes the section 
        <a
          href="/#"
        >
          link
        </a>
      </p>
    </div>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div>
        test settings content
      </div>
    </div>
  </div>
</div>
`;

exports[`<SettingsSubsection /> renders without description 1`] = `
<div>
  <div
    class="mx_SettingsSubsection"
  >
    <h3
      class="mx_Heading_h3 mx_SettingsSubsection_heading"
    >
      Test
    </h3>
    <div
      class="mx_SettingsSubsection_content"
    >
      <div>
        test settings content
      </div>
    </div>
  </div>
</div>
`;
